pkg/oauthflow/device.go

```diff
@@ -0,0 +1,157 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oauthflow
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"golang.org/x/oauth2"
+)
+
+const (
+	/* #nosec */
+	SigstoreDeviceURL = "https://oauth2.sigstore.dev/auth/device/code"
+	/* #nosec */
+	SigstoreTokenURL = "https://oauth2.sigstore.dev/auth/device/token"
+)
+
+type deviceResp struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationUri         string `json:"verification_uri"`
+	VerificationUriComplete string `json:"verification_uri_complete"`
+	Interval                int    `json:"interval"`
+}
+
+type tokenResp struct {
+	IDToken string `json:"id_token"`
+	Error   string `json:"error"`
+}
+
+type DeviceFlowTokenGetter struct {
+	MessagePrinter func(string)
+	Sleeper        func(time.Duration)
+	Issuer         string
+	CodeURL        string
+	TokenURL       string
+}
+
+func NewDeviceFlowTokenGetter(issuer, codeURL, tokenURL string) *DeviceFlowTokenGetter {
+	return &DeviceFlowTokenGetter{
+		MessagePrinter: func(s string) { fmt.Println(s) },
+		Sleeper:        time.Sleep,
+		Issuer:         issuer,
+		CodeURL:        codeURL,
+		TokenURL:       tokenURL,
+	}
+}
+
+func (d *DeviceFlowTokenGetter) deviceFlow(clientID string) (string, error) {
+	data := url.Values{
+		"client_id": []string{clientID},
+		"scope":     []string{"openid email"},
+	}
+
+	/* #nosec */
+	resp, err := http.PostForm(d.CodeURL, data)
+	if err != nil {
+		return "", err
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	parsed := deviceResp{}
+	if err := json.Unmarshal(b, &parsed); err != nil {
+		return "", err
+	}
+	d.MessagePrinter(fmt.Sprintf("Open your browser to: %s", parsed.VerificationUriComplete))
+
+	for {
+		// Some providers use a secret here, we don't need for sigstore oauth one so leave it off.
+		data := url.Values{
+			"grant_type":  []string{"urn:ietf:params:oauth:grant-type:device_code"},
+			"device_code": []string{parsed.DeviceCode},
+			"scope":       []string{"openid", "email"},
+		}
+
+		/* #nosec */
+		resp, err := http.PostForm(d.TokenURL, data)
+		if err != nil {
+			return "", err
+		}
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		tr := tokenResp{}
+		if err := json.Unmarshal(b, &tr); err != nil {
+			return "", err
+		}
+
+		if tr.IDToken != "" {
+			d.MessagePrinter("Token received!")
+			return tr.IDToken, nil
+		}
+		switch tr.Error {
+		case "access_denied", "expired_token":
+			return "", fmt.Errorf("error obtaining token: %s", tr.Error)
+		case "authorization_pending":
+		case "slow_down":
+		default:
```

Comment:
note that the error codes from https://tools.ietf.org/html/rfc6749#section-5.2 are still valid and we would end up in a loop repeating when we shouldn't.

need to move Sleeper calls up to cover the last two cases but not this one,  since there is no implicit fall through in golang.

Reply:
I think I got it.